### PR TITLE
Fix a bug where `INTERNAL_ERROR` is returned when gRPC message length is exceeded

### LINKED ISSUE #5818: Incorrect grpc status code on exceeding max response limit
There seems to be an issue with the way gRPC status codes are handled when the max response limit is exceeded.The following line:
https://github.com/line/armeria/blob/c208353b2edd94df20a1523d7b34bb349c781292/grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java#L153

could throw `AmeriaStatusRuntimeException` when the max response limit is set. The response future is then completed with that exception directly, bypassing the exception handler. This is different from exceptions occurring within the handler, which are transformed into StatusRuntimeException in the `close` method:

https://github.com/line/armeria/blob/74f6e3c301330708723e7e9a8bb960d07f1a7159/grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java#L216-L221.

As a result, the expected gRPC status code `RESOURCE_EXHAUSTED` is not being set when the response message exceeds the limit. Instead, I observe that the status code returned is `INTERNAL_ERROR`.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

If a request or response message exceeds the maximum allowed message length, `RESOURCE_EXHAUSTED` status should be returned to the client.

In `UnaryServerCall`, an exception raised `ArmeriaMessageFramer` is not handled by `GrpcExceptionHandlerFunction`. As a result, the exception is directly passed to the HTTP level `ServerErrorHandler` and a proper gRPC status is not delivered to the client.

While I was fixing the bug, I found a lot of code that calls `generateMetadataFromThrowable()` and then `fromThrowable()`. As the same pattern was used repeatedly, I also refactored `GrpcExceptionHandlerFunctionUtil` to return `StatusAndMetata` instead.

Modifications:

- Apply `GrpcExceptionHandlerFunction` when an exception is raised in `UnaryServerCall.doClose()`
- Refactor `GrpcExceptionHandlerFunctionUtil` and its calling site.
  - Add `InternalGrpcExceptionHandler` to wrap `GrpcExceptionHandlerFunction`.
- Remove the incorrect `@Nullable` annotation for `GrpcExceptionHandlerFunction`.
- Remove `ServerStatusAndMetadata.setResponseContent()` and use `RESPONSE_CONTENT` property for simpliciy.

Result:

- `GrpcService` now correctly returns `RESOURCE_EXHAUSTED` when a response message exceeds the maximum allowed length.
- Closes #5818

### CODE REVIEW

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java [hunk 1]
```diff
@@ -174,11 +175,19 @@ public void doClose(ServerStatusAndMetadata statusAndMetadata) {
...
+            final StatusAndMetadata statusAndMetadata0 = exceptionHandler().handle(ctx, ex);
+            final Status status = statusAndMetadata0.status();
+            final Metadata metadata = statusAndMetadata0.metadata();
+            assert metadata != null;
+            statusAndMetadata = new ServerStatusAndMetadata(status, metadata, true);
+
+            final ResponseHeadersBuilder trailersBuilder = defaultResponseHeaders().toBuilder();
+            final HttpResponse response = HttpResponse.of(
+                    (ResponseHeaders) statusToTrailers(ctx, trailersBuilder, status, metadata));
+            resFuture.complete(response);
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:190]: This block is the actual fix for the bug.

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java [hunk 2]
```diff
@@ -174,11 +175,18 @@ public void doClose(ServerStatusAndMetadata statusAndMetadata) {
 
             // Set responseContent before closing stream to use responseCause in error handling
             ctx.logBuilder().responseContent(GrpcLogUtil.rpcResponse(statusAndMetadata, responseMessage), null);
-            statusAndMetadata.setResponseContent(false);
             resFuture.complete(response);
         } catch (Exception ex) {
-            statusAndMetadata.shouldCancel();
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java:180]: So no need to call `shouldCancel` anymore?

File: grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/ServerStatusAndMetadata.java [hunk 3]
```diff
@@ -27,43 +27,30 @@
...
 
-    public boolean isShouldCancel() {
+    public boolean shouldCancel() {
         return shouldCancel;
     }
 
     /**
      * Tries to mark whether the call should be cancelled. If a call path
      * has already set the status to be cancelled, subsequent calls have no effect.
      */
```
Comment (jrhee17) [inline grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/ServerStatusAndMetadata.java:47]: nit; probably just removing the javadoc will be less confusing

File: grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java [hunk 4]
```diff
@@ -302,7 +289,7 @@ protected final void closeListener(ServerStatusAndMetadata statusAndMetadata) {
                 ctx.logBuilder().requestContent(GrpcLogUtil.rpcRequest(method, simpleMethodName), null);
             }
 
-            if (setResponseContent) {
+            if (!ctx.log().isAvailable(RequestLogProperty.RESPONSE_CONTENT)) {
```
Comment (jrhee17) [inline grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java:292]: 👍

**Review by jrhee17 — COMMENTED** [review]
Left a question but mostly looks good 👍

File: grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/InternalGrpcExceptionHandler.java [hunk 5]
```diff
@@ -24,29 +25,44 @@
...
-                                               Status status, Throwable cause, Metadata metadata) {
+    public Status handle(RequestContext ctx, Status status, Throwable cause, Metadata metadata) {
         final Throwable peeled = peelAndUnwrap(cause);
-        status = exceptionHandler.apply(ctx, status, peeled, metadata);
+        return handle0(ctx, status, peeled, metadata);
+    }
+
+    private Status handle0(RequestContext ctx, Status status, Throwable cause, Metadata metadata) {
+        if (status == Status.UNKNOWN) {
+            // If ArmeriaStatusException is thrown, it is converted to UNKNOWN and passed through close(Status).
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/InternalGrpcExceptionHandler.java:57]: This condition will be reached if a user calls `.onError(new ArmeriaStatusException(Code.RESOURCE_EXHAUSTED)` in a stub method.
- The upstream gRPC converts the exception into a status via `Status.fromThrowable()`.
- `UNKNOWN` is returned by `.fromThrowable()` and the `status.getCause()` is `ArmeriaStatusException`.
- The `cause` is passed to `InternalGrpcExceptionHandler`. https://github.com/line/armeria/blob/cc94d2f7280a106d89b925d7d0f6002f3f2a4b65/grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java#L211-L217
- `peelAndUnwrap()` converts `ArmeriaStatusException` to a `StatusRuntimeException` whose status is `RESOURCE_EXHAUSTED`.

Although it is covered by `GrpcMessageLengthTest`, let me add a dedicated test for it.

[the remark this answers is not in the document]
Comment (jrhee17) [reply grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/InternalGrpcExceptionHandler.java:58]: I see, I missed this point. Thanks for the explanation 👍

**Review by jrhee17 — APPROVED** [review]
👍 👍 👍

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

**Comment by ikhoon** [comment]
I found a leak in the CI reports and sent a separate PR (#5833) because it is unrelated to this PR.

**Pull request merged**